plugins/telink-dfu/fu-telink-dfu-hid-device.c

```diff
@@ -0,0 +1,409 @@
+#define FU_TELINK_DFU_HID_DEVICE_OTA_LENGTH	65
+#define FU_TELINK_DFU_HID_DEVICE_OTA_START_LEN	2
+#define FU_TELINK_DFU_HID_DEVICE_OTA_END_LEN	6
+#define FU_TELINK_DFU_HID_DEVICE_OTA_DATA_LEN	20
+#define FU_TELINK_DFU_HID_DEVICE_REPORT_ID	6
+#define HID_IFACE				2
+#define HID_EP_IN				(0x80 | 4)
+#define HID_EP_OUT				(0x00 | 5)
+#define HID_TOOL_VER(a, b)			((a) * 100 + (b))
+
+static gboolean
+fu_telink_dfu_hid_device_setup(FuDevice *device, GError **error)
+{
+	/* HidDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_telink_dfu_hid_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static FuStructTelinkDfuHidPkt *
+fu_telink_dfu_hid_device_create_packet(FuTelinkDfuCmd cmd,
+				       const guint8 *buf,
+				       gsize bufsz,
+				       GError **error)
+{
+	FuStructTelinkDfuHidPkt *pkt = fu_struct_telink_dfu_hid_pkt_new();
+	FuStructTelinkDfuHidPkt *payload = fu_struct_telink_dfu_hid_pkt_payload_new();
+	guint16 ota_data_len;
+
+	switch (cmd) {
+	case FU_TELINK_DFU_CMD_OTA_FW_VERSION:
+		ota_data_len = 0;
+		break;
+	case FU_TELINK_DFU_CMD_OTA_START:
+		ota_data_len = FU_TELINK_DFU_HID_DEVICE_OTA_START_LEN;
+		break;
+	case FU_TELINK_DFU_CMD_OTA_END:
+		ota_data_len = FU_TELINK_DFU_HID_DEVICE_OTA_END_LEN;
+		break;
+	default:
+		/* OTA data */
+		ota_data_len = FU_TELINK_DFU_HID_DEVICE_OTA_DATA_LEN;
+		break;
+	}
+
+	fu_struct_telink_dfu_hid_pkt_payload_set_ota_cmd(payload, cmd);
+	if (buf != NULL)
+		fu_struct_telink_dfu_hid_pkt_payload_set_ota_data(payload, buf, bufsz, NULL);
+	/* exclude the ota_cmd field */
+	fu_struct_telink_dfu_hid_pkt_payload_set_crc(
+	    payload,
+	    ~fu_crc16(FU_CRC_KIND_B16_USB, payload->data, payload->len - 2));
+	fu_struct_telink_dfu_hid_pkt_set_ota_data_len(pkt, ota_data_len);
+	fu_struct_telink_dfu_hid_pkt_set_payload(pkt, payload, error);
+	return pkt;
+}
+
+static gboolean
+fu_telink_dfu_hid_device_write(FuTelinkDfuHidDevice *self, guint8 *buf, gsize bufsz, GError **error)
+{
+	FuHidDevice *hid_dev = FU_HID_DEVICE(self);
+	FuHidDeviceFlags set_report_flag = FU_HID_DEVICE_FLAG_NONE;
+
+	if (self->hid_tool_ver >= HID_TOOL_VER(5, 2)) {
+		set_report_flag = FU_HID_DEVICE_FLAG_USE_INTERRUPT_TRANSFER;
+	}
+
+	if (!fu_hid_device_set_report(hid_dev,
+				      FU_TELINK_DFU_HID_DEVICE_REPORT_ID,
+				      buf,
+				      FU_TELINK_DFU_HID_DEVICE_OTA_LENGTH,
+				      FU_TELINK_DFU_HID_DEVICE_REPORT_TIMEOUT,
+				      set_report_flag,
+				      error)) {
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_telink_dfu_hid_device_write_blocks(FuTelinkDfuHidDevice *self,
+				      FuChunkArray *chunks,
+				      FuProgress *progress,
+				      GError **error)
+{
+	FuStructTelinkDfuHidLongPkt *long_pkt = fu_struct_telink_dfu_hid_long_pkt_new();
+	guint payload_index;
+	guint i;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, fu_chunk_array_length(chunks));
+	for (i = 0; i < fu_chunk_array_length(chunks); i++) {
```

Comment:
move `guint i` here pls.

Reply:
resolved in commit 0fb74d587e8aedf6a6bccabff38623ea034d224d